Evaluate, ngraph_reader check and Python nGraph API for GatherElements

pavel-esir opened the pull request:
Added Evaluate, Python nGraph API and ngraph_reader_check for GatherElements

ilyachur commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,82 @@
+#include <cassert>
+#include <numeric>
+
+//#include "ngraph/coordinate_transform.hpp"
```
Redundant header

pavel-esir replied:
fixed.

ilyachur commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,82 @@
+                }
+
+                int64_t count = 1;
+                for (int64_t i = 0; i < indices_shape.size(); i++)
```
Can we use `ngraph::shape_size()` instead of this cycle?

pavel-esir replied:
Thanks! replaced it with `ngraph::shape_size`

pavel-esir commented on ngraph/core/include/ngraph/op/gather_elements.hpp:
```diff
@@ -0,0 +1,55 @@
+
+                void validate_and_infer_types() override;
+                bool visit_attributes(AttributeVisitor& visitor) override;
+                virtual std::shared_ptr<Node>
```
removed `virtual` in op declaration PR #3467

pavel-esir commented:
Rebased on the latest master

rkazants commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,79 @@
+                for (int64_t i = 0; i < ngraph::shape_size(indices_shape); i++)
+                {
+                    data_idx = i - axis_mul * (((i / axis_mul) % data_shape[axis]) - indices[i]);
+                    out[i] = data[data_idx];
```
please check that indices are not out of allowed range and handle such invalid cases.

pavel-esir replied:
numpy will check itself if we exceed boundaries and throw error if it is the case. So I think adding additional check here will be redundant.

rkazants commented on ngraph/python/src/ngraph/opset6/ops.py:
```diff
@@ -0,0 +1,81 @@
+
+    @param data:       N-D tensor with data for gathering
+    @param indices:    N-D tensor with indices by which data is gathered
+    @param axis: axis along which elements are gathered
+    @return: The new node which performs GatherElements
```
nit: please fix indentation for description

pavel-esir replied:
Fixed

rkazants commented on ngraph/python/tests/test_ngraph/test_data_movement.py:
```diff
@@ -213,3 +213,18 @@ def test_gather_nd():
     assert node.get_output_element_type(0) == Type.f32
+
+
+def test_gather_eleVments():
```
typo: elements

pavel-esir replied:
Thanks. Fixed

rkazants commented on ngraph/test/backend/gather_elements.in.cpp:
```diff
@@ -0,0 +1,174 @@
+    auto gather_el = make_shared<op::v6::GatherElements>(arg1, arg2, axis);
+    auto fun = make_shared<Function>(OutputVector{gather_el}, ParameterVector{arg1, arg2});
+
+    std::vector<int32_t> data{1, 2, 3, 4};
+    std::vector<int32_t> indices{0, 1, 0, 0};
```
nit: please align tensor with defined shapes for better visual check.

pavel-esir replied:
Unfortunately `style-check` does not allow me to do that.

rkazants commented on ngraph/test/backend/gather_elements.in.cpp:
```diff
@@ -0,0 +1,174 @@
+    test_case1.add_input<int32_t>(indices);
+    test_case1.add_expected_output<float_t>(vector<float_t>{4.0f, 8.0f, 3.0f, 7.0f, 2.0f, 3.0f});
+    test_case1.run();
+}
```
please add case when indices are out of range

pavel-esir replied:
Added checks and unit-tests for that case.

rkazants reviewed (changes requested):
see comments

tomdol commented on ngraph/test/backend/gather_elements.in.cpp:
```diff
@@ -0,0 +1,174 @@
+    auto test_case1 = test::TestCase<TestEngine>(fun);
+    test_case1.add_multiple_inputs<int32_t>({data, indices});
+    test_case1.add_expected_output<int32_t>(vector<int32_t>{1, 4, 1, 2});
+    test_case1.run();
```
Just a suggestion to split this test into 3 individual tests (you can use parametrization, or put the code in a reusable function) because the first failing call of the `run()` method will stop the execution of the test. Thus the following checks will not be executed.

pavel-esir replied:
Agree. Splitted this test

pavel-esir replied:
@tomdol please review changes.

ilyachur commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,79 @@
+                if (axis < 0)
+                {
+                    axis += data_shape.size();
+                }
```
What if `axis` still less than 0?

pavel-esir replied:
If rank of `data` is static then axis have been checked to be in the acceptable range. If rank of `data` is dynamic should we add similar checks here as well?

ilyachur replied:
I can pass any value for axis in this method. I propose that at least we should add an assert in order to check that axis is not less than 0. In other case (without the assert) we can get a crash in the future.

pavel-esir replied:
Agreed. Added check for axis values. Please review

tomdol commented on ngraph/test/runtime/ie/unit_test.manifest:
```diff
@@ -1540,3 +1540,27 @@ onnx_controlflow_loop_infinite
 IE_GPU.range_v4_trunc_inputs
+
+# not implemented jet on CPU and GPU plugins
+IE_CPU.evaluate_1D_gather_elements_3_indices_int32
```
If you remove the `IE_*.` prefix and leave the test name only in this file, such line will disable this test for both plugins.

pavel-esir replied:
On GPU GatherElements will be implemented much after CPU so it's better to keep this separation. But thanks for the information, I didn't know that.

lazarevevgeny commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,93 @@
+                int64_t data_idx;
+                for (int64_t i = 0; i < ngraph::shape_size(indices_shape); i++)
+                {
+                    data_idx = i - axis_mul * (((i / axis_mul) % data_shape[axis]) - indices[i]);
```
To be honest, I don't like this approach because it involves division and modulo which are quite expensive operations.

@vladislav-volkov , @ilyachur , can we use here indices iterators which have been optimized recently instead of this hand-crafted iteration?

ilyachur replied:
@vladislav-volkov Could you comment?
If I remember right we cannot use the optimized ConstTransformIterator.

Also we discussed with @pavel-esir and he propose to add an optimised implementation for 2d tensors it should cover majority of our cases.

vladislav-volkov replied:
I suggest looking at PR 2897 and PR 3633. I hope we can somehow reuse the same approach.

pavel-esir replied:
Implemented without Iterators as discussed personally with @vladislav-volkov.

lazarevevgeny commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,93 @@
+                }
+
+                int64_t data_idx;
+                for (int64_t i = 0; i < ngraph::shape_size(indices_shape); i++)
```
`ngraph::shape_size(indices_shape)` as well as `ngraph::shape_size(data_shape)` can be computed once before the `for` loop.

pavel-esir replied:
I suspect compiler optimizes these statements, `indices_shape` and `data_shape` are defined as constants in signature.

lazarevevgeny replied:
They are constant, but the check in the `for` loop is performed before each iteration. I assume that the compiler is smart enough, but we can avoid relying on it by simply adding new variable. Anyway, let's wait for the resolution of comment above, because it may change everything.

pavel-esir replied:
Agreed and moved `ngraph::shape_size` outside of `for` loop.

lazarevevgeny commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <algorithm>
+#include <cassert>
```
Do we really need `cassert` library?

pavel-esir replied:
I will remove them.

vladislav-volkov commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,125 @@
+
+                // in 6D case with axis = 2
+                // N5*N4*N3
+                size_t max_inner_sum = ngraph::shape_size(
```
Here is better to write overload for ngraph::shape_size which can operate with two iterators.

vladislav-volkov replied:
Yes, you are right (:

pavel-esir replied:
I was not able to simply override `ngraph::shape_size` so leaved size calculations in for loops inside gather_elements.

vladislav-volkov commented on ngraph/core/reference/include/ngraph/runtime/reference/gather_elements.hpp:
```diff
@@ -0,0 +1,125 @@
+                    out[i] = data[data_idx];
+
+                    i++;
+                    if (i == outer_threshold)
+                    {
+                        outer_sum++;
+                        outer_threshold += outer_threshold_inc;
+                    }
+
+                    inner_sum++;
+                    if (inner_sum == max_inner_sum)
+                    {
+                        inner_sum = 0;
+                    }
```
Also it would be better to implement explicitly inner and outer loops? We can exclude if-s from loop.

pavel-esir replied:
Done. Please review

pavel-esir commented:
Added separate calculation for 2D case to speed up calculation since it's most frequent case. Also rewritten solution with 3 nested `for` loops: outer, middle and inner. @vladislav-volkov please review
